tensorflow/java/src/gen/cc/java_defs.h

```diff
@@ -0,0 +1,266 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_JAVA_SRC_GEN_CC_JAVA_DEFS_H_
+#define TENSORFLOW_JAVA_SRC_GEN_CC_JAVA_DEFS_H_
+
+#include <string>
+#include <vector>
+#include <deque>
+
+#include "tensorflow/core/platform/env.h"
+
+namespace tensorflow {
+namespace java {
+
+/// \brief An enumeration of different modifiers commonly used in Java
+enum Modifier {
+  PUBLIC    = (1 << 0),
+  PROTECTED = (1 << 1),
+  PRIVATE   = (1 << 2),
+  STATIC    = (1 << 3),
+  FINAL     = (1 << 4),
+};
+
+class Annotation;
+
+/// \brief A definition of any kind of Java type (classes, interfaces...)
+///
+/// Note that most of the data fields of this class are only useful in specific
+/// contexts and are not required in many cases. For example, annotations and
+/// supertypes are only useful when declaring a type.
+class Type {
+ public:
+  enum Kind {
+    PRIMITIVE, CLASS, INTERFACE, ENUM, GENERIC, ANNOTATION
+  };
+  struct Comparator {
+    bool operator() (const Type& type1, const Type& type2) {
+      return type1.name_ < type2.name_ || type1.package_ < type2.package_;
+    }
+  };
+  static Type Primitive(const string& name) {
+    return Type(Type::PRIMITIVE, name, "");
+  }
+  static Type Class(const string& name, const string& package = "") {
+    return Type(Type::CLASS, name, package);
+  }
+  static Type Interface(const string& name, const string& package = "") {
+    return Type(Type::INTERFACE, name, package);
+  }
+  static Type Enum(const string& name, const string& package = "") {
+    return Type(Type::ENUM, name, package);
+  }
+  static Type Generic(const string& name) {
+    return Type(Type::GENERIC, name, "");
+  }
+  static Type Wildcard() {
+    return Type(Type::GENERIC, "", "");
+  }
+  static Type ClassOf(const Type& type) {
+    return Class("Class").param(type);
+  }
+  static Type ListOf(const Type& type) {
+    return Interface("List", "java.util").param(type);
+  }
+  static Type IterableOf(const Type& type) {
+    return Interface("Iterable").param(type);
+  }
+  const Kind& kind() const { return kind_; }
+  const string& name() const { return name_; }
+  const string& package() const { return package_; }
+  const string& descr() const { return descr_; }
+  Type& descr(const string& descr) { descr_ = descr; return *this; }
+  const std::vector<Type>& params() const { return params_; }
+  Type& param(const Type& param) {
```

Comment:
Let's call this `add_param` (similarly for `add_annotation` and `add_supertype`). That makes it clearer at the call site that these are lists that are added to (as opposed to say `descr()`, which replaces any existing value)